### ISSUE TITLE
[Bug] find_homography_dlt_iterated sometimes fails

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#795]
## 🐛 Bug

The functionality `find_homography_dlt_iterated` sometimes causes gradient check to crash.
For this reason the tests related to gradient check have been temporally marked as xfail:
- https://github.com/kornia/kornia/blob/master/test/geometry/test_homography.py#L81

**Comment by dkoguciuk** [comment]
Hi @ducha-aiki ,

I can try to help with this, can you point me how to reproduce the bug?

Daniel

**Comment by dkoguciuk** [comment]
Hi @edgarriba , @ducha-aiki ,

## The conclusion

Following pytorch documentation on SVD [HERE](https://pytorch.org/docs/stable/generated/torch.svd.html): `Gradients computed using U and V may be unstable if the input is not full rank or has non-unique singular values.` And I believe this could be the problem here: **for some inputs gradients are unstable**.

## My methodology

I've fixed the torch seed while creating random input tensors and switch it back to the original seed just afterward. Here's the code snippet:

```
    @pytest.mark.grad
    def test_gradcheck(self, device, dtype):

        #######################################################################
        # HELP VARS
        manual_seed = 42
        iterations = 20
        #######################################################################

        #######################################################################
        # SEED
        initial_seed = torch.random.initial_seed()
        torch.manual_seed(manual_seed)
        #######################################################################

        points_src = torch.rand(1, 10, 2, device=device, dtype=torch.float64, requires_grad=True)
        points_dst = torch.rand_like(points_src)

        #######################################################################
        # SEED
        torch.manual_seed(initial_seed)
        #######################################################################

        weights = torch.ones_like(points_src)[..., 0]
        assert gradcheck(find_homography_dlt_iterated, (points_src, points_dst, weights, 3.0, iterations),
                         rtol=1e-6, atol=1e-6, raise_exception=True)
```

And for some seeds (42 for my environment) the test passes regardless of the number of iterations, but for other seed values (5 for my environment) the test is passed for a small number of iterations (2-5), but for more iterations (20-100), it fails. So the conclusion could be just like I've pointed out above.

## Possible solution

When we test functionality with `torch.SVD` we could test gradients for a couple of seeds before throwing an exception. Here's a possible pseudocode:

```
    @pytest.mark.grad
    def test_gradcheck(self, device, dtype):

        # Save initial seed
        initial_seed = torch.random.initial_seed()
        max_number_of_checks = 5

        # Test gradients for a couple of times
        current_seed = initial_seed
        for i in range(max_number_of_checks-1):
            torch.manual_seed(current_seed)
            points_src = torch.rand(1, 10, 2, device=device, dtype=torch.float64, requires_grad=True)
            points_dst = torch.rand_like(points_src)
            weights = torch.ones_like(points_src)[..., 0]
            success = gradcheck(find_homography_dlt_iterated, (points_src, points_dst, weights, 3.0, iterations),
                rtol=1e-6, atol=1e-6, raise_exception=False)
            if success:
                torch.manual_seed(initial_seed)
                return
            else:
                current_seed = here_randomly_choose_new_seed

        # All iterations failed
        assert gradcheck(find_homography_dlt_iterated, (points_src, points_dst, weights, 3.0, iterations),
                rtol=1e-6, atol=1e-6, raise_exception=True)
```

**Comment by ducha-aiki** [comment]
Wow, that is an impressive investigation! Your solution looks good for me as well. 
@edgarriba you opinion?

Also, it would be great, if pytorch incorporated robust SVD implementation from this Fua paper
https://arxiv.org/pdf/2104.03821

Do we have any influence over that?

**Comment by shijianjian** [comment]
BTW, @dkoguciuk non-full rank matrix error is also happening with torch.inverse as well. There might also need a fix as well. For now, I used torch.pinverse. Do you have any insights for this?

**Comment by dkoguciuk** [comment]
Hi @ducha-aiki , Hi @shijianjian ,

> Do we have any influence over that?

I certainly don't :smile: But as you can see [HERE](https://github.com/pytorch/pytorch/issues/49886#issuecomment-819644142) it was already mentioned there, so maybe someday it will be incorporated into PyTorch.

> Do you have any insights for this?

It is hard for me to say because I am no linear algebra expert by any means, but I can give you some intuitions. I assume there could be applications where using real inverse is preferred. Maybe we should let the user decide which SVD method he wants to use (the regular or the robust one) or if he wants to use `torch.inverse` or `torch.pinverse`? But that will complicate the API a lot.

This "solution" that I've listed above is not solving any problem with SVD or inverse -- only helping to perform the gradcheck better.

**Comment by edgarriba** [comment]
If `torch.pinverse` solves the issue then I'd go for it

**Comment by dkoguciuk** [comment]
@edgarriba that is not the case here -- I suspect `find_homography_dlt` sometimes fails, because of `torch.svd` (as described above) and we don't have `torch.psvd` just yet. What do you think about the workaround I've proposed above? Can I code it and propose a PR?

**Comment by edgarriba** [comment]
Yes, hi ahead

**Comment by dkoguciuk** [comment]
Fixed by #1022 and #1032 .
